### ISSUE TITLE
[JENKINS-6822] Don't log exception when unable to find memory mon impl

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-23#8314]
<p>Don't log the exception here, if you want to keep it, then log it as a DEBUG, so that normal operation on a system that doesn't support this monitor doesn't always puke up this stack trace.</p>

```
jvm 1    | WARN  [Monitoring thread for Free Swap Space started on Mon Jun 21 14:35:06 PDT 2010] h.n.SwapSpaceMonitor$1 - Failed to monitor master for Free Swap Space
jvm 1    | java.io.IOException: No suitable implementation found
jvm 1    | 	at org.jvnet.hudson.MemoryMonitor.obtain(MemoryMonitor.java:60) ~[memory-monitor-1.2.jar:na]
jvm 1    | 	at org.jvnet.hudson.MemoryMonitor.get(MemoryMonitor.java:31) ~[memory-monitor-1.2.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$MonitorTask.call(SwapSpaceMonitor.java:103) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$MonitorTask.call(SwapSpaceMonitor.java:99) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.remoting.LocalChannel.call(LocalChannel.java:45) ~[remoting-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$1.monitor(SwapSpaceMonitor.java:83) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$1.monitor(SwapSpaceMonitor.java:81) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:200) ~[hudson-core-1.363.jar:na]

```

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=user57">user57</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-6822" target="_blank">Don't log exception when unable to find memory mon impl</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2010-08-02T10:12:53+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 0
<li><b>imported</b>: 2025-11-23
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Don't log the exception here, if you want to keep it, then log it as a DEBUG, so that normal operation on a system that doesn't support this monitor doesn't always puke up this stack trace.</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>jvm 1    | WARN  [Monitoring thread for Free Swap Space started on Mon Jun 21 14:35:06 PDT 2010] h.n.SwapSpaceMonitor$1 - Failed to monitor master for Free Swap Space
jvm 1    | java.io.IOException: No suitable implementation found
jvm 1    | 	at org.jvnet.hudson.MemoryMonitor.obtain(MemoryMonitor.java:60) ~[memory-monitor-1.2.jar:na]
jvm 1    | 	at org.jvnet.hudson.MemoryMonitor.get(MemoryMonitor.java:31) ~[memory-monitor-1.2.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$MonitorTask.call(SwapSpaceMonitor.java:103) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$MonitorTask.call(SwapSpaceMonitor.java:99) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.remoting.LocalChannel.call(LocalChannel.java:45) ~[remoting-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$1.monitor(SwapSpaceMonitor.java:83) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.SwapSpaceMonitor$1.monitor(SwapSpaceMonitor.java:81) ~[hudson-core-1.363.jar:na]
jvm 1    | 	at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:200) ~[hudson-core-1.363.jar:na]
</pre>
</div></div></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>andrzejdoyle:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-6822?focusedId=139575&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-139575">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm in two minds about this, personally.  I understand exactly why you're asking for it, since by the looks of things this will <b>always</b> happen on Windows machines.  The thing is, there are two different classes of scenario when this exception will occur:</p>

<p>1) On a system where Hudson doesn't know how to monitor memory, and so will always run without this functionality (e.g. Windows).  Here this should be logged at a DEBUG level as you say (or perhaps even not logged at all, merely noted in the documentation).</p>

<p>2) On a system where we do expect to be able to monitor the memory, but initialisation failed for some reason.  In this case we should arguably be logging at an ERROR level.</p>

<p>Currently it's not possible to tell between the two cases; the initialisation step is simply a case of trying all known MemoryMonitors in turn, and throwing the above exception if none of them can be started successfully.  An ability to distinguish between a true <b>failure</b> of the monitor startup, and simply an environment where it knows it cannot start, would allow this difference in error reporting here.  As it is, your proposed solution means that if the memory monitor fails to start on Linux (or some other environment where one would expect it to work), there would be nothing in the logs to indicate why.</p>

<p>Thus I think hypothetically, the cleanest solution would be to add an isSupported() method to the MemoryMonitor interface, and call this on each implementation in turn.  If no implementations are possibly supported, just return null and carry on (logging this at DEBUG/INFO as it's expected here).  And of course, if implementations are supported but fail, throw an exception to indicate failure as happens now.</p>

<p>I don't know how feasible it is to separate out the preconditions like this; currently the implementations simply plough ahead and expect the required environment to exist.  On the plus side, we can be conservative when implementing this method; false positives are not the end of the world and will simply end up in an identical situation to now.  Consequently filtering out <b>any</b> environment as not supported is an improvement of some sorts.</p>

<p>Any thoughts on this?</p>
</pre>
</details>

</ul>
</details>

<p>I'm in two minds about this, personally.  I understand exactly why you're asking for it, since by the looks of things this will <b>always</b> happen on Windows machines.  The thing is, there are two different classes of scenario when this exception will occur:</p>

<p>1) On a system where Hudson doesn't know how to monitor memory, and so will always run without this functionality (e.g. Windows).  Here this should be logged at a DEBUG level as you say (or perhaps even not logged at all, merely noted in the documentation).</p>

<p>2) On a system where we do expect to be able to monitor the memory, but initialisation failed for some reason.  In this case we should arguably be logging at an ERROR level.</p>

<p>Currently it's not possible to tell between the two cases; the initialisation step is simply a case of trying all known MemoryMonitors in turn, and throwing the above exception if none of them can be started successfully.  An ability to distinguish between a true <b>failure</b> of the monitor startup, and simply an environment where it knows it cannot start, would allow this difference in error reporting here.  As it is, your proposed solution means that if the memory monitor fails to start on Linux (or some other environment where one would expect it to work), there would be nothing in the logs to indicate why.</p>

<p>Thus I think hypothetically, the cleanest solution would be to add an isSupported() method to the MemoryMonitor interface, and call this on each implementation in turn.  If no implementations are possibly supported, just return null and carry on (logging this at DEBUG/INFO as it's expected here).  And of course, if implementations are supported but fail, throw an exception to indicate failure as happens now.</p>

<p>I don't know how feasible it is to separate out the preconditions like this; currently the implementations simply plough ahead and expect the required environment to exist.  On the plus side, we can be conservative when implementing this method; false positives are not the end of the world and will simply end up in an identical situation to now.  Consequently filtering out <b>any</b> environment as not supported is an improvement of some sorts.</p>

<p>Any thoughts on this?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-6822?focusedId=140667&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-140667">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Also note that as per <a href="https://issues.jenkins.io/browse/JENKINS-2194" title="Failed to monitor Free Swap Space" class="issue-link" data-issue-key="JENKINS-2194"><del>JENKINS-2194</del></a>, this exception is reported only once per slave, which I think is a reasonable thing to do.</p>

<p>As I mentioned above, for the platforms that Hudson doesn't support the monitoring, please file a separate ticket so that we can expand the support. </p>
</pre>
</details>

</ul>
</details>

<p>Also note that as per <a href="https://issues.jenkins.io/browse/JENKINS-2194" title="Failed to monitor Free Swap Space" class="issue-link" data-issue-key="JENKINS-2194"><del>JENKINS-2194</del></a>, this exception is reported only once per slave, which I think is a reasonable thing to do.</p>

<p>As I mentioned above, for the platforms that Hudson doesn't support the monitoring, please file a separate ticket so that we can expand the support. </p>
